lib/logstash/devutils/rspec/shared_examples.rb

```diff
@@ -0,0 +1,23 @@
+require 'rspec/wait'
+
+RSpec.shared_examples "an interruptible input plugin" do
+  describe "#stop" do
+    let(:queue) { SizedQueue.new(20) }
+    subject do
+      s = described_class.new(config)
+      s.register
+      s
+    end
+    it "returns from run" do
+      consumer_thread = Thread.new(queue) { |queue| loop { queue.pop } }
+      plugin_thread = Thread.new(subject, queue) { |subject, queue| subject.run(queue) }
+      # the run method is a long lived one, so it should still be running after "a bit"
+      sleep 0.5
```

Comment:
Me and Guy talked about this and the big issue is that in a lot of situations it's not possible to reliably implement this when using client libraries.
In the example I gave before with a blocking subscribe method such as 

```
def run(queue)
  s = rabbitmq_server.new(@options)
  s.subscribe do |event|
    queue << LogStash::Event.new(event)
  end
end
```
The `subscribe` action can actually take a few round trips to the server, setting up the exchange, queue, and binding between them, even more with SSL.

Marking the plugin ready after `new` means that it's possible to
1. start logstash
2. `rabbitmq_server.new(@options)`
3. set `ready?` to true
4. start subscribe
5. user does ctrl+c
6. `stop` is called, checks that ready? is true and does a stop action for a blocking subscribe
7. action above does nothing because subscribe is not blocked yet
6. subscribe finishes setting up and blocks
7. rabbitmq ready for events!
8. never quits